File: day1/main.go
package main

import (
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

func main() {
	argv := os.Args[1:]
	if len(argv) < 1 {
		fmt.Printf("Syntax: day1 [input.txt]\n")
		return
	}

	rawInput, err := os.ReadFile(argv[0])
	if err != nil {
		panic(err)
	}
	lines := strings.Split(strings.ReplaceAll(string(rawInput), "\r\n", "\n"), "\n")

	listL := make([]int, 0, len(lines))
	listR := make([]int, 0, len(lines))
	countR := make(map[int]int)

	for i := 0; i < len(lines); i++ {
		if len(lines[i]) < 1 {
			continue
		}
		rawNums := strings.Fields(lines[i])
		if len(rawNums) < 2 {
			continue
		}

		l, err := strconv.Atoi(rawNums[0])
		if err != nil {
			panic(err)
		}
		listL = append(listL, l)

		r, err := strconv.Atoi(rawNums[1])
		if err != nil {
			panic(err)
		}
		listR = append(listR, r)
		countR[r]++
	}

	slices.Sort(listL)
	slices.Sort(listR)

	totalDistance := 0
	similarityScore := 0
	for i := 0; i < len(listL); i++ {
		l := listL[i]
		r := listR[i]

		diff := l - r
		if diff < 0 {
			diff = -diff
		}
		totalDistance += diff

		similarityScore += l * countR[l]
	}

	fmt.Printf("Total Distance  : %v\n", totalDistance)
	fmt.Printf("Similarity Score: %v\n", similarityScore)
}

File: day1/main.cpp
#include <algorithm>
#include <fstream>
#include <map>
#include <sstream>
#include <string>
#include <vector>

int main(int argc, char * argv[])
{
    if (argc < 2) {
        printf("Syntax: main [input.txt]\n");
        return 1;
    }
    const std::string inputFilename = argv[1];

    std::vector<int> listL;
    std::vector<int> listR;
    std::map<int, int> countR;

    std::ifstream f;
    f.open(inputFilename);
    std::stringstream ss;
    ss << f.rdbuf();
    const std::string contents = ss.str();

    std::size_t lineStart = 0;
    for (;;) {
        const std::size_t lineEnd = contents.find("\n", lineStart);
        if (lineEnd == std::string::npos) {
            break;
        }

        const std::string line = contents.substr(lineStart, lineEnd - lineStart);
        if (line.size() > 0) {
            int l = 0;
            int r = 0;
            std::istringstream(line) >> l >> r;
            listL.push_back(l);
            listR.push_back(r);

            std::map<int, int>::iterator e = countR.find(r);
            if (e == countR.end()) {
                countR.insert(std::pair<int, int>(r, 1));
            } else {
                ++e->second;
            }
        }

        lineStart = lineEnd + 1;
    }

    std::sort(listL.begin(), listL.end());
    std::sort(listR.begin(), listR.end());

    int totalDistance = 0;
    int similarityScore = 0;
    for (int i = 0; i < listL.size(); ++i) {
        int l = listL[i];
        int r = listR[i];
        int diff = l - r;
        if (diff < 1) {
            diff *= -1;
        }
        totalDistance += diff;

        std::map<int, int>::iterator e = countR.find(l);
        if (e != countR.end()) {
            similarityScore += l * e->second;
        }
    }

    printf("Total Distance  : %d\n", totalDistance);
    printf("Similarity Score: %d\n", similarityScore);
    return 0;
}
